In Multiplayer docs, added property and RPC types #1165 (#2081)
o3de/o3de.org: content/docs/user-guide/networking/multiplayer/overview.md

- **Authority-to-Server**: For handling host migration.

- **Autonomous-to-Authority**: For gathering information about client metrics, such as monitoring the health of clients.

For networking properties that replicate from Authority, a replication hierarchy applies. The replicate-to rules trickle up the hierarchy in the following way: An Authority-to-Client replication also replicates to Autonomous and Server roles. An Authority-to-Autonomous replication also replicates to the Server role. Finally, an Authority-to-Server replication only replicates to the Server. This behavioral hierarchy ensures that if the Authority ever migrates to the other server, then the other server has the right property information.

## Remote procedure calls (RPCs)

Developers can use *remote procedure calls* in O3DE to invoke a function on a remote endpoint. An RPC is a useful mechanism for signaling and notifying about events across networked endpoints. Unlike network properties, a developer chooses when to invoke an RPC. RPCs are not guaranteed to arrive in the order they are sent. O3DE offers both reliable and unreliable RPCs. By default, RPCs are reliable. 

> **Reliable** RPCs use a queue to guarantee the delivery of a message. When sending any reliable packet, the packet is also inserted into a priority queue and given a timeout value that's related to the latency of the connection that the message is sent on. On timeout, if the reliable packet was not explicitly acknowledged, the packet will be retransmitted. On the receiving end, the O3DE client tracks every reliable packet received and guarantees that any packet will only be delivered at most once. While this feature provides *guaranteed* delivery, it doesn't provide *ordered* delivery.

Reviewer: Suggested text:
On the receiving end, the O3DE client tracks every reliable packet received and guarantees that any packet will be delivered only once.

**Unreliable** RPCs are sent over a "fire and forget" method. The host sending the message has no way to ensure that the message was received.

### Types of RPCs

Similar to network properties, RPCs have two important fields: `InvokeFrom` and `HandleOn`. These fields describe which role the RPC is invoked from and which role it's handled on.

| Field | Description | Values | 
| - | - | - |
| `InvokeFrom` | Tells the network which role invokes this RPC. | Server, Autonomous, Client | 
| `HandleOn` | Tells the network which roles handles this RPC. If Authority, the RPC is handled by the server that has authority over that entity. If Autonomous, the RPC is handled only by the relevant player's local client. If Client, the RPC is handled on all clients. | Authority, Autonomous, Client | 

The following are four types of RPCs with a valid "invoked from and handled on" relationship and their possible use cases:

> - **Authority-to-Autonomous**: For example, sending corrections about the game state to the user.

Reviewer: Suggested text:
For example, sending corrections about the game state to the user.
to
In an example use case, it sends corrections about the game state to the user.

> - **Authority-to-Client**: Authority sends calls to all the clients. For example, sending information about particle effects. 

Reviewer: Suggested text:
Authority sends calls to all the clients. For example, sending information about particle effects. 
to
Authority sends calls to all clients. For example, it sends information about particle effects.

- **Server-to-Authority**: This is required to communicate information between entities. For example, suppose in a multi-server setup, EntityA is owned by ServerA and EntityB is owned by ServerB. If EntityA communicates directly to EntityB, EntityA will be talking to a proxy, not the real EntityB. Server-to-Authority ensures that messages always find the entity with authority. When a player wants to deal damage, Autonomous informs the Server, and the Server sends the DealDamage function to Authority. 

> - **Autonomous-to-Authority**: Used for sending user settings information that affects user input and is used during input-process time rather than input-creation time, such as mouse sensitivity and input controls. 

Reviewer: Suggested text:
Used for sending user settings information...
to
Sends user settings information...

After, the AzNetworking layer becomes aware that the world was updated. This occurs automatically through ACK vector replication, and does not require any further action from the client.

### Types of network properties

Network properties have two important fields: `ReplicateFrom` and `ReplicateTo`. 

> Together, these define which role can replicate to which role. You can only replicate property values *from* Authority and Autonomous roles. Under the server-authority model of O3DE, this ensures that you never accidentally replicate from an unprivileged client. 

Reviewer 2: Actually this is about only replicating from roles that have write access to component data. If you only have read access (no controller), then you can't change any values, therefore there could never actually be anything for that role to replicate.

Author: ```suggestion
Together, these define which role can replicate to which role. You can only replicate property values *from* Authority and Autonomous roles. A `ReplicateFrom` Authority relationship creates a server-authority model, which ensures that you never accidentally replicate from an unprivileged client. 
```

Reviewer 3: That suggestion looks good. It may be good to note that ReplicateFrom Autonomous can be useful for sending non-authoritative data like telemetry.

> Properties can be replicated *to* any role, since all active clients in the session must share information and some may not necessarily need to be used in the server-authoritative model.

Reviewer 4: Should it maybe be changed to `all participants in the session`?

```suggestion
Properties can be replicated *to* any role, since all participants in the session may need information from any other participant.
```

| Field | Description | Values | 
| - | - | - |
| `ReplicateFrom` | Tells the network which role can send information about the changes that were made to this property. | Authority, Autonomous | 

> | `ReplicateTo` | Tells the network which role can receive information about the changes that were made to this property. | Server, Authority, Autonomous, Client | 

Reviewer 2: These abbreviated descriptions are perfect 👍

> The following are four types of networking properties with a valid "from and to" relationship that you can define, and their possible use cases:

Author: I haven't always found the docs to be clear about the user's entry point to the system, so I do tend to over-index on it. Understandable if you think it sounds unnatural here!

- **Authority-to-Client**: For handling client, or "simulated", properties.

- **Authority-to-Autonomous**: For handling autonomous-only properties.

- **Authority-to-Server**: For handling host migration.

- **Autonomous-to-Authority**: For gathering information about client metrics, such as monitoring the health of clients.

> For networking properties that replicate from Authority, a replication hierarchy applies. The replicate-to rules trickle up the hierarchy in the following way: An Authority-to-Client replication also replicates to Autonomous and Server roles. An Authority-to-Autonomous replication also replicates to the Server role. Finally, an Authority-to-Server replication only replicates to the Server. This behavioral hierarchy ensures that if the Authority ever migrates to the other server, then the other server has the right property information.

Reviewer 2: Perfect 👍

## Remote procedure calls (RPCs)

> Remote procedure calls are how O3DE allows for pushing messages between hosts, rather than sending an update that's based on the server world state. O3DE offers both *reliable* and *unreliable* RPCs. By default, RPCs are reliable.

Author: @lmbr-pip  I'd certainly like to improve this - do you estimate that its a quick fix (I can do in this PR) or shall I create an issue for us to improve this at a separate time?

> Remote procedure calls are how O3DE allows for pushing messages between hosts, rather than sending an update that's based on the server world state. O3DE offers both *reliable* and *unreliable* RPCs. By default, RPCs are reliable.

Reviewer 5: Suggestion:
_Remote procedure calls are how O3DE allows for pushing messages between multiplayer endpoints, rather than sending an update that's based on the server world state (ie: network properties). O3DE offers both *reliable* and *unreliable* RPCs. By default, RPCs are reliable. Note: RPCs aren’t guaranteed to arrive in the order they are sent._

> Remote procedure calls are how O3DE allows for pushing messages between hosts, rather than sending an update that's based on the server world state. O3DE offers both *reliable* and *unreliable* RPCs. By default, RPCs are reliable.

Reviewer 5: Echo @AMZN-puvvadar with the addition that _RPCs aren’t guaranteed to arrive in the order they are sent._

> Remote procedure calls are how O3DE allows for pushing messages between hosts, rather than sending an update that's based on the server world state. O3DE offers both *reliable* and *unreliable* RPCs. By default, RPCs are reliable.

Author: Thank you all for the explanation - I'll get this in

**Unreliable** RPCs are sent over a "fire and forget" method. The host sending the message has no way to ensure that the message was received.

### Types of RPCs

Similar to network properties, RPCs have two important fields: `InvokeFrom` and `HandleOn`. These fields describe which role the RPC is invoked from and which role it's handled on.

| Field | Description | Values | 
| - | - | - |
| `InvokeFrom` | Tells the network which role invokes this RPC. | Server, Autonomous, Client | 
| `HandleOn` | Tells the network which roles handles this RPC. If Authority, the RPC is handled by the server that has authority over that entity. If Autonomous, the RPC is handled only by the relevant player's local client. If Client, the RPC is handled on all clients. | Authority, Autonomous, Client | 

The following are four types of RPCs with a valid "invoked from and handled on" relationship that you can define, and their possible use cases:

- **Authority-to-Autonomous**: For example, sending corrections about the game state to the user.

- **Authority-to-Client**: Authority sends calls to all the clients. For example, sending information about particle effects. 

> - **Server-to-Authority**: This is required to communicate information between entities. For example, suppose in a multi-server setup, EntityA is owned by ServerA and EntityB is owned by ServerB. If EntityA communicates directly to EntityB, EntityA will be talking to a proxy, not the real EntityB. Server-to-Authority ensures that messages always find the entity with authority. when a player wants to deal damage, Autonomous informs the Server, and the Server sends the DealDamage function to Authority. 

Reviewer 5: Capital “When”

Author: ```suggestion
- **Server-to-Authority**: This is required to communicate information between entities. For example, suppose in a multi-server setup, EntityA is owned by ServerA and EntityB is owned by ServerB. If EntityA communicates directly to EntityB, EntityA will be talking to a proxy, not the real EntityB. Server-to-Authority ensures that messages always find the entity with authority. When a player wants to deal damage, Autonomous informs the Server, and the Server sends the DealDamage function to Authority. 
```
